noxfile.py

```diff
@@ -4,52 +4,104 @@
 
 import os
 import pathlib
-from itertools import chain
-from typing import TYPE_CHECKING, Callable, Dict, List, Tuple, TypeVar
+import sys
+from typing import Any, Dict, List, Optional, Sequence, Tuple
 
 import nox
 
-if TYPE_CHECKING:
-    from typing_extensions import Concatenate, ParamSpec
-
-    P = ParamSpec("P")
-    T = TypeVar("T")
-
-    NoxSessionFunc = Callable[Concatenate[nox.Session, P], T]
-
-
 # see https://pdm-project.org/latest/usage/advanced/#use-nox-as-the-runner
 os.environ.update(
     {
         "PDM_IGNORE_SAVED_PYTHON": "1",
     },
 )
+nox.needs_version = ">=2025.5.1"
+
+PYPROJECT = nox.project.load_toml()
+
+
+def use_min_python_of(python: str, *, preferred: Optional[str] = None) -> str | None:
+    """Use the minimum necessary python for this run, but if the environment is a specific venv, then use that one."""
+    major, minor = python.split(".")
+    if sys.version_info < (int(major), int(minor)):
+        # If the current Python version is less than the required version, use the required version
+        return preferred or python
+    return None
 
 
 nox.options.error_on_external_run = True
-nox.options.reuse_existing_virtualenvs = True
+nox.options.reuse_venv = "yes"
 nox.options.sessions = [
     "lint",
     "check-manifest",
     "slotscheck",
     "pyright",
     "test",
 ]
-nox.needs_version = ">=2022.1.7"
 
 
 # used to reset cached coverage data once for the first test run only
 reset_coverage = True
 
 
+def install_deps(
+    session: nox.Session,
+    *,
+    extras: Sequence[str] = (),
+    groups: Sequence[str] = (),
+    dependencies: Sequence[str] = (),  # a parameter itself for pip
+    project: bool = True,
+) -> None:
+    """Helper to install dependencies from a group, using uv if venv_backend is uv."""
+    if not project and extras:
+        raise TypeError("Cannot install extras without also installing the project")
+
+    command: List[str] = []
+
+    # install with UV
+    command = [
+        "pdm",
+        "sync",
+        "--fail-fast",
+        "--only-keep",
+    ]
+
+    # see https://pdm-project.org/latest/usage/advanced/#use-nox-as-the-runner
+    env: Dict[str, Any] = {
+        "PDM_IGNORE_SAVED_PYTHON": "1",
+    }
```

Comment:
This is already being set above, and nox should already be automatically passing environment variables from the invoking process to the child